File: sdk/android/story-sdk/src/main/kotlin/com/openstory/sdk/internal/network/SdkFeedResponseNormalizer.kt
package com.openstory.sdk.internal.network

import java.net.URI

internal object SdkFeedResponseNormalizer {
    fun normalizeLoopbackUrls(
        response: SdkFeedResponsePayload,
        baseUrl: String,
    ): SdkFeedResponsePayload {
        val baseUri = runCatching { URI(baseUrl) }.getOrNull() ?: return response
        val baseScheme = baseUri.scheme ?: return response
        val baseHost = baseUri.host ?: return response
        val basePort = if (baseUri.port == -1) {
            defaultPort(baseScheme)
        } else {
            baseUri.port
        }

        fun rewrite(url: String): String {
            val uri = runCatching { URI(url) }.getOrNull() ?: return url
            val host = uri.host?.lowercase() ?: return url
            if (host != LOOPBACK_HOST && host != LOOPBACK_IP) {
                return url
            }

            return URI(
                baseScheme,
                uri.userInfo,
                baseHost,
                basePort,
                uri.path,
                uri.query,
                uri.fragment,
            ).toString()
        }

        return response.copy(
            resolvedSet = response.resolvedSet?.copy(
                groups = response.resolvedSet.groups.map { group ->
                    group.copy(
                        logoUrl = rewrite(group.logoUrl),
                        stories = group.stories.map { story ->
                            story.copy(
                                asset = story.asset.copy(url = rewrite(story.asset.url)),
                                posterAsset = story.posterAsset?.copy(url = rewrite(story.posterAsset.url)),
                            )
                        },
                    )
                },
            ),
        )
    }

    private fun defaultPort(scheme: String): Int {
        return when (scheme.lowercase()) {
            "http" -> 80
            "https" -> 443
            else -> -1
        }
    }

    private const val LOOPBACK_HOST = "localhost"
    private const val LOOPBACK_IP = "127.0.0.1"
}

File: sdk/ios/Sources/OpenStorySDK/Internal/Network/SdkFeedResponseNormalizer.swift
import Foundation

internal enum SdkFeedResponseNormalizer {
    static func normalizeLoopbackURLs(
        response: SdkFeedResponsePayload,
        baseURL: String
    ) -> SdkFeedResponsePayload {
        guard
            let baseComponents = URLComponents(string: baseURL),
            let baseScheme = baseComponents.scheme,
            let baseHost = baseComponents.host
        else {
            return response
        }

        let basePort = baseComponents.port ?? defaultPort(for: baseScheme)

        func rewrite(_ value: String) -> String {
            guard var components = URLComponents(string: value) else {
                return value
            }

            let host = components.host?.lowercased() ?? ""
            guard host == "localhost" || host == "127.0.0.1" else {
                return value
            }

            components.scheme = baseScheme
            components.host = baseHost
            components.port = basePort
            return components.string ?? value
        }

        let rewrittenGroups = response.resolvedSet?.groups.map { group in
            SdkFeedGroupPayload(
                id: group.id,
                revisionId: group.revisionId,
                title: group.title,
                bottomLabel: group.bottomLabel,
                logoURL: rewrite(group.logoURL),
                badge: group.badge,
                stories: group.stories.map { story in
                    SdkFeedStoryPayload(
                        id: story.id,
                        revisionId: story.revisionId,
                        title: story.title,
                        mediaType: story.mediaType,
                        imageDurationMs: story.imageDurationMs,
                        asset: SdkFeedAssetPayload(
                            id: story.asset.id,
                            url: rewrite(story.asset.url),
                            mimeType: story.asset.mimeType,
                            width: story.asset.width,
                            height: story.asset.height,
                            durationMs: story.asset.durationMs
                        ),
                        posterAsset: story.posterAsset.map {
                            SdkFeedAssetPayload(
                                id: $0.id,
                                url: rewrite($0.url),
                                mimeType: $0.mimeType,
                                width: $0.width,
                                height: $0.height,
                                durationMs: $0.durationMs
                            )
                        },
                        cta: story.cta
                    )
                }
            )
        }

        return SdkFeedResponsePayload(
            clientId: response.clientId,
            placementKey: response.placementKey,
            context: response.context,
            resolvedSet: response.resolvedSet.map {
                SdkFeedSetPayload(
                    id: $0.id,
                    revisionId: $0.revisionId,
                    placementKey: $0.placementKey,
                    isFallback: $0.isFallback,
                    groups: rewrittenGroups ?? []
                )
            },
            generatedAt: response.generatedAt
        )
    }

    private static func defaultPort(for scheme: String) -> Int? {
        switch scheme.lowercased() {
        case "http":
            return 80
        case "https":
            return 443
        default:
            return nil
        }
    }
}
